(feat-docs): Adding docs for nfs kill chaos experiment (#258)
litmuschaos/litmus-docs: docs/openebs-nfs-kill-chaos.md

---

> id: openebs-nfs-kill-chaos

Reviewer: same name as chart

title: OpenEBS NFS KILL Chaos Experiment Details
sidebar_label: OpenEBS NFS Kill Chaos
---
------

## Experiment Metadata

<table>
  <tr>
    <th> Type </th>
    <th> Description </th>
    <th> Tested K8s Platform </th>
  </tr>
  <tr>
    <td> OpenEBS </td>

>     <td> Kill the OpenEBS NFS provisioner container and check if pods mounted with RWM modes.  </td>

Reviewer: same description as chart/exp

    <td> GKE </td>
  </tr>
</table>

## Prerequisites

- Ensure that the Litmus Chaos Operator is running by executing `kubectl get pods` in operator namespace (typically, `litmus`). If not, install from [here](https://docs.litmuschaos.io/docs/getstarted/#install-litmus)

> - Ensure that the `openebs-nfs-kill-chaos` experiment resource is available in the cluster by executing `kubectl get chaosexperiments` in the `openebs` namespace. If not, install from [here](https://hub.litmuschaos.io/charts/openebs/experiments/openebs-nfs-kill-chaos)

Reviewer: name

Reviewer: link for exp

> - The DATA_PERSISTENCE is used when `EXTERNAL_APP_CHECK` sets to true and it is enabled by provide the application's info in a configmap volume so that the experiment can perform necessary checks. By Default it will using `Busybox` data persistency check.

Reviewer: The "DATA_PERSISTENCE"env variable takes effect only if the "EXTERNAL_APP_CHECK" is enabled. By default, a stateful busybox deployment is used to create and validate data persistence on the RMW NFS persistent volumes.

> Busybox data persistence check create a configmap as shown below in the application namespace (replace with actual credentials):

Reviewer: The data persistence also needs the users to create a minimal configmap upfront (like described below) in the namespace where the experiment resources are created.

```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: openebs-nfs-kill-cm
data:
  parameters.yml: | 
    blocksize: 4k
    blockcount: 1024
    testfile: exampleFile
```

## Entry Criteria

> - OpenEBS NFS pods are healthy before chaos injection

Reviewer: ```suggestion
- OpenEBS NFS provisioner pod(s) are healthy after chaos injection

> - Pods using NFS RWM mode are healthy before chaos injection

Reviewer: ```suggestion
- Pods using NFS PVs in RWM mode are healthy before chaos injection
```

## Exit Criteria

> - OpenEBS NFS pods are healthy after chaos injection

Reviewer: ```suggestion
- OpenEBS NFS provisioner are healthy after chaos injection
```

> - Pods using NFS RWM mode are healthy after chaos injection

Reviewer: ```suggestion
-  Pods using NFS PVs in RWM mode are healthy before chaos injection
```

## Details

> - This scenario validates graceful & forced terminations of OpenEBS NFS container

Reviewer: ```suggestion
- This scenario validates crashes/failure of OpenEBS NFS provisioner container
```

  resources: ["pods","pods/exec","pods/log", "deployments","events","jobs","configmaps","services","persistentvolumeclaims","storageclasses","persistentvolumes","chaosexperiments","chaosresults","chaosengines"]
  verbs: ["create","list","get","patch","update","delete"]
---
apiVersion: rbac.authorization.k8s.io/v1beta1
kind: ClusterRoleBinding
metadata:
  name: nfs-kill-sa
  namespace: default
  labels:
    name: nfs-kill-sa
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: nfs-kill-sa
subjects:
- kind: ServiceAccount
  name: nfs-kill-sa
  namespace: default
```

### Prepare ChaosEngine

- Provide the application info in `spec.appinfo`
- Override the experiment tunables if desired in `experiments.spec.components.env`
- To understand the values to provided in a ChaosEngine specification, refer [ChaosEngine Concepts](chaosengine-concepts.md)

#### Supported Experiment Tunables

<table>
  <tr>
    <th> Variables </th>
    <th> Description  </th>
    <th> Specify In ChaosEngine </th>
    <th> Notes </th>
  </tr>
  <tr>
    <td> TARGET_CONTAINER  </td>
    <td> The container to be killed inside the pod </td>
    <td> Mandatory </td>

>     <td> If the TARGET_CONTAINER is not provided it will delete the first container </td>

Reviewer: ```suggestion
    <td> If the TARGET_CONTAINER is not provided it will delete the first container specified/listed in the deployment spec</td>
```

  </tr>
  <tr>
    <td> LIB_IMAGE  </td>
    <td> The pumba/containerd image used to run the kill command </td>
    <td> Optional </td>
    <td> Defaults to `gaiaadm/pumba:0.6.5`, for containerd runtime use `gprasath/crictl:ci`</td>
  </tr>
  <tr>
    <td> LIB  </td>
    <td> The category of lib use to inject chaos </td>
    <td> Optional  </td>
    <td> It can be pumba or containerd </td>
  </tr>
  <tr>
    <td> RAMP_TIME </td>

>     <td> Period to wait before injection of chaos in sec </td>

Reviewer: ```suggestion
    <td> Period to wait before and after injection of chaos in sec </td>
```
